python/pyhdk/_builder.pyx

```diff
@@ -0,0 +1,722 @@
+
+      for lhs_col in lhs_cols:
+        if not isinstance(lhs_col, str):
+          raise TypeError("Only strings are allowed in lhs_cols. Provided: {type(lhs_col)}.")
+        lhs_col_names.push_back(lhs_col)
+
+      if rhs_cols is None:
+        res.c_node = self.c_node.joinByCols((<QueryNode>rhs_node).c_node, lhs_col_names, how)
+      else:
+        if isinstance(rhs_cols, str) or not isinstance(rhs_cols, Iterable):
+          rhs_cols = [rhs_cols]
+
+        if len(lhs_cols) != len(rhs_cols):
+          raise ValueError("Mismatch length of 'lhs_cols' and 'rhs_cols'.")
+
+        for rhs_col in rhs_cols:
+          if not isinstance(rhs_col, str):
+            raise TypeError("Only strings are allowed in rhs_cols. Provided: {type(rhs_col)}.")
+          rhs_col_names.push_back(rhs_col)
+
+        res.c_node = self.c_node.joinByColPairs((<QueryNode>rhs_node).c_node, lhs_col_names, rhs_col_names, how)
+
+    res._hdk = self._hdk
+    return res
+
+
+  def _process_expr_arg(self, expr):
+    if isinstance(expr, (int, str)):
+      return self.ref(expr)
+    elif isinstance(expr, QueryExpr):
+      return expr
+    else:
+      raise TypeError(f"Unexpected query expression. Expected int, str or QueryExpr. Provided: {type(expr)}")
+
+  def _process_agg_expr_arg(self, expr):
+    if isinstance(expr, str):
+      res = QueryExpr()
+      res.c_expr = self.c_node.parseAggString(expr)
+      return res
+    elif isinstance(expr, QueryExpr):
+      return expr
+    else:
+      raise TypeError(f"Unexpected aggregate query expression. Expected str or QueryExpr. Provided: {type(expr)}")
+
+  def _process_sort_field_arg(self, field, key=None):
+    sort_dir = "asc"
+    null_pos = "last"
+
+    if key is None:
+      if isinstance(field, tuple):
+        if len(field) < 1 or len(field) > 3:
+          raise TypeError(f"Sort field tuple should have 1 to 3 elements but got {len(field)}.")
+        expr = self._process_expr_arg(field[0])
+        if len(field) > 1:
+          sort_dir = field[1]
+        if len(field) > 2:
+          null_pos = field[2]
+      else:
+        expr = self._process_expr_arg(field)
+    else:
+      expr = self._process_expr_arg(key)
+      if isinstance(field, tuple):
+        if len(field) < 1 or len(field) > 2:
+          raise TypeError(f"Sort direction tuple should have 1 to 2 elements but got {len(field)}.")
+        sort_dir = field[0]
+        if len(field) > 1:
+          null_pos = field[1]
+      else:
+        sort_dir = field
+
+    if not isinstance(sort_dir, str):
+      raise TypeError(f"Sort direction should be a string. Provided: {sort_dir}.")
+
+    if not isinstance(null_pos, str):
+      raise TypeError(f"Nulls position should be a string. Provided: {null_pos}.")
+
+    return expr, sort_dir, null_pos
+
+  @property
+  def is_scan(self):
+    return self.c_node.node().get().isNode[CScan]()
+
+  @property
+  def size(self):
+    res = self.c_node.size()
+    # For scans we hide virtual column
```

Comment:
I didn't follow closely, `c_node.size()` isn't this a call to C++ API? And if so, doesn't it already contain the logic for subtracting `1` in case of a scan?

Reply:
In C++ we actually do not hide the virtual column in such a manner. So we do all these modifications for virtual columns in Python.

Reply:
Do you think this should be the user's responsibility? I'd think it's a rather specific feature for modin, isn't it?

Reply:
Well, `rowid` virtual column is useful and is used by Modin. I don't want users to 'see' this column when they ask for node size or use an integer index to choose a column. I wrote some examples and it feels really confusing when this virtual column affects size and indexing (e.g. `ht[-1]` would give you the last column in most cases but for scans, it would be the `rowid` column). I believe the current Python implementation is the right way to go (we still can reference `rowid` column by its name). Probably, we should adjust C++ API implementation to have a match.

Reply:
Yeah, I think consistent semantics is important. Anyway, it's not related to this PR. LGTM!